File: 2020/11_SeatingSystem/conway.lua
-- ======================================================================
-- Seating System
--   Advent of Code 2020 Day 11 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         c o n w a y . l u a
-- ======================================================================
-- A solver for the Advent of Code 2020 Day 11 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Conway = { part2 = false, text = {}, seats = {}, leave = 4 }

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local SEAT_FLOOR = "."
local SEAT_EMPTY = "L"
local SEAT_OCCUPIED = '#'
local SEAT_BEYOND = ' '

local NEIGHBORS = { {-1, -1}, {-1, 0}, {-1, 1},
                    { 0, -1},          { 0, 1},
                    { 1, -1}, { 1, 0}, { 1, 1} }

-- ======================================================================
--                                                                 Conway
-- ======================================================================

-- Object for Seating System

function Conway:Conway (o)

  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.seats = {}
  if o.part2 then
    o.leave = 5
  else  
    o.leave = 4
  end
  
  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Conway:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)

  -- 1. Loop for each line of the text
  for _, line in ipairs(text) do

    -- 2. Split the line into individual characters (seats and floors)
    local row = {}
    line:gsub(".", function(c) table.insert(row, c) end)
  
    -- 3. Add the row to the current status
    table.insert(self.seats, row)
  end

end

function Conway:count_occupied() 
  -- Count the currently occupied seats
  
  -- 1. Start with nothing
  local result = 0
  
  -- 2. Loop for all of the rows
  for _, row in ipairs(self.seats) do
    
    -- 3. Loop for all of the seats (and some floors) in the row
    for _, seat in ipairs(row) do
      
      -- 4. If the seat is occupied, increment the count
      if seat == SEAT_OCCUPIED then
        result = result + 1
      end
    end
  end
  
  -- 5. Return the count of occupied seats
  return result
end

function Conway:seat(row, col)
  -- Return the status of a single seat
  return self.seats[row][col]
end

function Conway:neighbor(row, col, delta)
  -- Return the status of a seat near the specified seat
  
  -- 1. Compute the seat number
  local delta_row = row + delta[1]
  local delta_col = col + delta[2]
  
  -- 2. If off the seating plan, return floor
  if delta_row < 1 or delta_col < 1 or delta_row > #self.seats or delta_col > #self.seats then
    return SEAT_BEYOND
  end
  
  -- 3. Return the seat status
  return self:seat(delta_row, delta_col)
end

  
function Conway:neighbors(row, col)
  -- Return the number of occupied seats near the specified seat
  
  -- 1. Start with nothing
  local num_occupied = 0
  
  -- 2. Loop for all of the neighbors
  for _, delta in ipairs(NEIGHBORS) do
    
    -- 3. Get the status of the nearby seat
    local nbor = self:neighbor(row, col, delta)
    -- print(row, col, delta[1], delta[2], nbor)
    
    -- 4. If the seat is occupied, increment occupied count
    if nbor == SEAT_OCCUPIED then
      num_occupied = num_occupied + 1
      
    -- 5. If part2 we look until the next seat  
    elseif self.part2 and nbor == SEAT_FLOOR then
      nbor = self:sight_line_neighbor(row, col, delta)
      if nbor == SEAT_OCCUPIED then
        num_occupied = num_occupied + 1
      end
    end
  end
  
  -- 6. Return the number of nearby occupied seats
  return num_occupied
end

function Conway:sight_line_neighbor(row, col, delta)
  -- Returns the next seat in the sight line
  
  -- 1. Start with with floor
  local nbor = SEAT_FLOOR
  local delta_row = row + delta[1]
  local delta_col = col + delta[2]
  
  -- 2. Loop until we are not at a a floor space
  while nbor == SEAT_FLOOR do
    
    -- 3. Check the next space
    nbor = self:neighbor(delta_row, delta_col, delta)
    
    -- 4. Advance
    delta_row = delta_row + delta[1]
    delta_col = delta_col + delta[2]
  end
  
  -- 5. Return the non-Floor value
  return nbor
end
   
  
function Conway:next_round()
  -- Advance seats occupation, returns True is something changed
  
  -- 1. Start with nothing
  local future = {}
  local changed = false
  
  -- 2. Loop for all of the rows and seats in the row
  for rnum, row in ipairs(self.seats) do
    local future_row = {}
    for cnum, seat in ipairs(row) do
      local future_seat = seat
      
      -- 3. Get the number of occupied seats
      local num_occupied = 0
      if seat ~= SEAT_FLOOR then 
        num_occupied = self:neighbors(rnum, cnum)
      end
            
      -- 4. An empty seat becomes occupied if there are no nearby occupied seats
      if seat == SEAT_EMPTY and num_occupied == 0 then
        future_seat = SEAT_OCCUPIED
        changed = true
      -- 5. An occupied seat becomed empty if four or more seats adjacent to it are occupied  
      elseif seat == SEAT_OCCUPIED and num_occupied >= self.leave then
        future_seat = SEAT_EMPTY
        changed = true
      end
      
      -- 6. Set the future seat in the future row
      table.insert(future_row, future_seat)
    end
    -- 7. Set the future row 
    table.insert(future, future_row)
  end
  
  -- 8. The future is now
  self.seats = future
  
  -- 9. Return true if the occupation of a seat changed
  return changed
end

function Conway:show(before, after)
  -- Show the seats
  
  -- 1. If there is a before label, print it
  if before ~= nil then
    print(before)
  end
  
  -- 2. Loop for all of the rows
  for _, row in ipairs(self.seats) do
    
    -- 3. Print the row as text
    print(table.concat(row, ''))
  end

  -- 4. If there is an after label, print it
  if after ~= nil then
    print(after)
  end
end
  
function Conway:until_stable()
  -- Run multiple rounds until stable, return number of occupied seats
  
  -- 1. Loop until no changes
  while self:next_round() do end
  
  -- 2. Return the number of occupied seats
  return self:count_occupied()
end


function Conway:part_one(args)
  -- Returns the solution for part one

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:until_stable()
end


function Conway:part_two(args)
  -- Returns the solution for part two

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part two
  return self:until_stable()
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Conway

-- ======================================================================
-- end                       c o n w a y . l u a                      end
-- ======================================================================

File: 2020/11_SeatingSystem/conway.ts
// ======================================================================
// Seating System
//   Advent of Code 2020 Day 11 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           c o n w a y . t s
//
// A solver for the Advent of Code 2020 Day 11 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type Seat = '.' | 'L' | '#' | ' ';
type SeatRow = Seat[];
type Seats = SeatRow[];

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const SEAT_FLOOR: Seat = '.';
const SEAT_EMPTY: Seat = 'L';
const SEAT_OCCUPIED: Seat = '#';
const SEAT_BEYOND: Seat = ' ';

const NEIGHBORS = [
  [-1, -1], [-1, 0], [-1, 1],
  [+0, -1], /*    */ [+0, 1],
  [+1, -1], [+1, 0], [+1, 1],
];

// ======================================================================
//                                                                 Conway
// ======================================================================

export class Conway {
  // Object for Seating System
  text: string[];

  part2: boolean;

  seats: Seats;

  leave: number;

  constructor(text: string[], part2 = false) {
    // Create a Conway object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.seats = [];
    this.leave = this.part2 ? 5 : 4;

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText();
    }
  }

  processText() {
    // Process the text into seats
    // 1. Loop for every row in the text
    for (let rindx = 0; rindx < this.text.length; rindx += 1) {
      const textRow = this.text[rindx];

      // 2. Start with an empty row
      const seatrow: SeatRow = [];

      // 3. Loop for every character in the text row
      for (let cindx = 0; cindx < textRow.length; cindx += 1) {
        let textCol = textRow.substr(cindx, 1) as Seat;
        if (textCol !== SEAT_FLOOR && textCol !== SEAT_EMPTY) {
          console.log(`Unexpected character ${textCol} at row ${rindx} column ${cindx}`);
          textCol = SEAT_FLOOR;
        }

        // 4. Add the seat text to the seat row
        seatrow.push(textCol);
      }
      // 5. Add the row of seats to seats
      this.seats.push(seatrow);
    }
  }

  countOccupied() {
    // Returns number of occupied seats

    // 1.Start with nothing
    let result = 0;

    // 2. Loop for all of the rows
    for (let rindx = 0; rindx < this.seats.length; rindx += 1) {
      const row = this.seats[rindx];

      // 3. Loop for all of the columns
      for (let cindx = 0; cindx < row.length; cindx += 1) {
        const col = row[cindx];

        // 4. If someone is in the chair, increment count
        if (col === SEAT_OCCUPIED) {
          result += 1;
        }
      }
    }

    // 5. Return the number of occupied seats
    return result;
  }

  seat(row: number, col: number): Seat {
    // Return the status of the specified seat
    return this.seats[row][col];
  }

  neighbor(row: number, col: number, delta: number[]): Seat {
    // Return the status of a nearby seat

    // 1. Compute the seat number
    const deltaRow = row + delta[0];
    const deltaCol = col + delta[1];

    // 2. If off the seating plane, return beyond
    if (deltaRow < 0 || deltaCol < 0
      || deltaRow >= this.seats.length || deltaCol >= this.seats.length) {
      return SEAT_BEYOND;
    }

    // 3. Return the seat status
    return this.seat(deltaRow, deltaCol);
  }

  neighbors(row: number, col: number): number {
    // Return number of occupied seats nearby

    // 1. Start with nothing
    let result = 0;

    // 2. Loop for all of the neighbors
    for (let nindx = 0; nindx < NEIGHBORS.length; nindx += 1) {
      const delta = NEIGHBORS[nindx];

      // 3. Get the status of the nearby seat
      const nbor = this.neighbor(row, col, delta);

      // 4. If the seat is occupied, increment count
      if (nbor === SEAT_OCCUPIED) {
        result += 1;
        // 5. If part two, we look until the next seat
      } else if (this.part2 && nbor === SEAT_FLOOR) {
        const slnbor = this.sightLineNeighbor(row, col, delta);
        if (slnbor === SEAT_OCCUPIED) {
          result += 1;
        }
      }
    }

    // 6. Return the number of occupied seats
    return result;
  }

  nextRound(): boolean {
    // Advance seat occupation, return true if something changed

    // 1. Start with nothing
    const future: Seats = [];
    let changed = false;

    // 2. Loop for all of the rows and the seats in the row
    for (let rindx = 0; rindx < this.seats.length; rindx += 1) {
      const currentRow = this.seats[rindx];
      const futureRow: SeatRow = [];
      for (let cindx = 0; cindx < currentRow.length; cindx += 1) {
        let futureSeat = currentRow[cindx];

        // 3. Get the number of occupied seats
        let numOccupied = 0;
        if (futureSeat !== SEAT_FLOOR) {
          numOccupied = this.neighbors(rindx, cindx);
        }

        // 4. An empty seat becomes occupied if there are no nearby occupied seats
        //    And an occupied seat become empty if there to many people nearby
        switch (futureSeat) {
          case SEAT_EMPTY:
            if (numOccupied === 0) {
              futureSeat = SEAT_OCCUPIED;
              changed = true;
            }
            break;
          case SEAT_OCCUPIED:
            if (numOccupied >= this.leave) {
              futureSeat = SEAT_EMPTY;
              changed = true;
            }
            break;
          default:
            break;
        }

        // 5. Set the future seat in the future row
        futureRow.push(futureSeat);
      }
      // 6. Set the future row
      future.push(futureRow);
    }
    // 7. The future is now
    this.seats = future;

    // 8. Return true is a seat occupation changed
    return changed;
  }

  untilStable(): number {
    // Return the number of seats occupied when things settle down

    // 1. Loop until there are no changes
    while (this.nextRound()) {
      // empty
    }

    // 2. Return the count of occupied seats
    return this.countOccupied();
  }

  sightLineNeighbor(row: number, col: number, delta: number[]): Seat {
    // Returns the next seat in the sight line

    // 1. Start with the floor
    let nbor = SEAT_FLOOR;
    let deltaRow = row + delta[0];
    let deltaCol = col + delta[1];

    // 2. Loop until we are not at a floor space
    while (nbor === SEAT_FLOOR) {
      // 3. Check the next space
      nbor = this.neighbor(deltaRow, deltaCol, delta);

      // 4. Advance
      deltaRow += delta[0];
      deltaCol += delta[1];
    }

    // 5. Return the non-floor value
    return nbor;
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    return this.untilStable();
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                      c o n w a y . t s                     end
// ======================================================================
